### PR TITLE
#3258 - Supplier Errors Part 3

### PR DESCRIPTION
**Pull request by guru-aot**
CAS errors are shown in the row expander of the CAS Supplier Information table.
image

E2E test case is created and existing is updated for the changed API.

Failure in the sonarcloud is due to code duplication but I am not able to find the duplication in the code submitted in this PR.

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts [hunk 1]
```diff
@@ -63,6 +66,67 @@ describe("CASSupplierAESTController(e2e)-getCASSuppliers", () => {
...
+    const student = savedCASSupplier1.student;
+    const savedCASSupplier2 = await saveFakeCASSupplier(
+      db,
+      { student },
+      {
+        initialValues: {
+          supplierStatus: SupplierStatus.ManualIntervention,
+          errors: [
+            "[0034] SIN is already in use.",
+            "[9999] Duplicate Supplier , Reason: [0065]- Possible duplicate exists, please use online form",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts:87]: Is it supposed to emulate a real CAS error?
The string does not look formated.

File: sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts [hunk 2]
```diff
@@ -63,6 +66,67 @@ describe("CASSupplierAESTController(e2e)-getCASSuppliers", () => {
...
+            supplierNumber: savedCASSupplier1.supplierNumber,
+            supplierProtected: savedCASSupplier1.supplierProtected,
+            supplierStatus: savedCASSupplier1.supplierStatus,
+            isValid: savedCASSupplier1.isValid,
+            supplierSiteCode:
+              savedCASSupplier1.supplierAddress.supplierSiteCode,
+            siteStatus: savedCASSupplier1.supplierAddress.status,
+            addressLine1: savedCASSupplier1.supplierAddress.addressLine1,
+            siteProtected: savedCASSupplier1.supplierAddress.siteProtected,
+            errors: null,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts:129]: Is this not the same scenario validated in the previous E2E?
If yes, I recommend removing it.

File: sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts (hunk 1 above)
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts:87]: its as per the response we got from the API
image

File: sources/packages/backend/libs/test-utils/src/factories/cas-supplier.ts [hunk 3]
```diff
@@ -63,6 +63,7 @@ export function createFakeCASSupplier(
   casSupplier.supplierStatus =
     options?.initialValues?.supplierStatus ??
     SupplierStatus.PendingSupplierVerification;
+  casSupplier.errors = options?.initialValues?.errors ?? null;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/cas-supplier.ts:66]: Is there a need to set null?

File: sources/packages/web/src/views/aest/student/CASSupplierInformation.vue [hunk 4]
```diff
@@ -58,6 +60,36 @@
...
+                isExpanded,
+                toggleExpand,
+              }"
+            >
+              <v-btn
+                v-if="item.errors && item.errors.length > 0"
+                :icon="
+                  isExpanded(internalItem)
+                    ? 'mdi-chevron-down'
+                    : 'mdi-chevron-up'
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/CASSupplierInformation.vue:87]: We have default icons in place for expander that can/must be reused, IMO.
```vue
<v-btn
  v-if="item.errors && item.errors.length > 0"
  :icon="
    isExpanded(internalItem)
      ? '$expanderExpandIcon'
      : '$expanderCollapseIcon'
  "
  variant="text"
  @click="toggleExpand(internalItem)"
></v-btn>
```

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

File: sources/packages/web/src/views/aest/student/CASSupplierInformation.vue
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/CASSupplierInformation.vue:1]: Please review the below AC:
```
Update the CAS table column order, add new columns, update column names
```

Comment (andrepestana-aot) [reply sources/packages/web/src/views/aest/student/CASSupplierInformation.vue:1]: Take a look at the order and column names:
image

File: sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts (hunk 2 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/cas-supplier/_tests_/cas-supplier.aest.controller.getCASSupplier.e2e-spec.ts:129]: The errors can be added to the previous test.

**Review by andrepestana-aot — COMMENTED** [review]
Good job. There's just some missing changes for the columns and minor comments.

File: sources/packages/web/src/views/aest/student/CASSupplierInformation.vue
Comment (guru-aot) [reply sources/packages/web/src/views/aest/student/CASSupplierInformation.vue:1]: image

**Comment by andrewsignori-aot** [comment]
> SonarCloud test are failing!

Please take a look at the PR description and if you have some idea, please feel free to share 😉

File: sources/packages/web/src/types/contracts/DataTableContract.ts [hunk 5]
```diff
@@ -320,45 +320,55 @@ export const CASSupplierInformationHeaders = [
...
-    key: "siteStatus",
+    key: "supplierProtected",
   },
   {
     title: "Site protected",
     sortable: false,
     key: "siteProtected",
   },
+  {
+    title: "Address line 1",
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/DataTableContract.ts:363]: The AC states "Address line" only.

**Review by andrepestana-aot — APPROVED** [review]
Looks good to me! Thanks for doing the changes!

File: sources/packages/web/src/types/contracts/DataTableContract.ts
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/DataTableContract.ts:1]: The table labels vary between starting the second-word uppercase/lowercase.
We should not have it inside the same table. Please check with biz which one should be.

**Review by andrewsignori-aot — APPROVED** [review]
Great job, the UI is looking great 👍

File: sources/packages/web/src/types/contracts/DataTableContract.ts
Comment (guru-aot) [reply sources/packages/web/src/types/contracts/DataTableContract.ts:1]: image
image
 Confirmed with @CarlyCotton

**Pull request merged**